### ISSUE TITLE
Validation of Istio annotations on K8s objects

### ISSUE THREAD

**Issue by esnible** [istio/istio#17272]
Users sometimes place Istio annotations in the wrong place.  For example, Pod annotations on the Deployment outside of the Template.

Users sometimes misspell annotations, or supply a value in the wrong format.

Istio's validators, probably the new framework validators from https://github.com/istio/istio/pull/16350 should issue warnings.  Possibly this could go into the existing webhook and sidecar injector, but I don't know.

If there is a problem it is unclear how to report it.  For 1.4 we can report it through `istioctl x analyze`.  I am not sure if we can add Status info to K8s objects like Deployment.

We should also check Istio annotations on Istio objects.

cc @linsun @ozevren

**Comment by davidebbo** [comment]
Yep, seems like a great scenario for which we could potentially write an analyzer. /cc @sushicw

**Comment by esnible** [comment]
@ozevren @sushicw I need your advice.  `istio.io/istio/galley/pkg/config/meta/metadata` provides collections for Kubernetes objects that Istio cares about: Endpoints, Namespaces, Nodes, Pods, Services, and Ingresses.

It doesn't know about Deployments at all.  This means I can't use the analysis framework to validate the use case Lin originally described to me: users putting Pod annotations on their Deployments instead of Deployment.Spec.Template.

One possibility is to add Deployments to the things already watched and validated.  That would be easy.  The other possibility is to follow the `ownerReferences` fields for the things already being validated so that the Pod validator can validate the owner.

**Comment by davidebbo** [comment]
I think adding adding Deployments makes sense. When I wrote my injection analyzer earlier (https://github.com/istio/istio/pull/17192), the framework didn't include Pods, so I had to add that.

But I'll let @ozevren or @sushicw confirm.

**Comment by ozevren** [comment]
You can add deployments to the localAnalysis snapshot in the metadata file.

https://github.com/istio/istio/blob/9d9ede26ae0fe970cee0bce42beaa030ca6405d8/galley/pkg/config/meta/metadata/metadata.yaml#L564

You may also need to update the runtime code to ensure these can be read in their native form (as opposed to as unstructureds):
https://github.com/istio/istio/blob/9d9ede26ae0fe970cee0bce42beaa030ca6405d8/galley/pkg/config/source/kube/rt/known.go#L41

This should allow you to read the deployments for istioctl, without polluting the Galley MCP serving path.

**Comment by geeknoid** [comment]
We've now updated annotations.yaml in istio/api to specify the set of resource types that each annotation supports. This file can be imported by the validator.

We'll also shortly have this added to the documentation on istio.io. I'm working on that right now.

**Comment by esnible** [comment]
@ozevren I added Deployments as you suggested.  It was more difficult than you suggested.  The Commit is https://github.com/istio/istio/pull/17714/commits/676f6d200d54e1ef2139eaa2e8a798ebdda90b8b 

Some of my difficulties may have been caused by leaving off the optional `group:` field in a .yaml due to not remembering that Deployments have a group but Pods do not.  This led me down the path of adding Deployments to a test.  Once I understood all the places that Deployment needed to be added it was not difficult to actually add it.
